### ISSUE TITLE
Interact with real backend

### ISSUE THREAD

**Issue by samselikoff** [samselikoff/ember-cli-mirage#31]
somehow

**Comment by thatikuntap** [comment]
how do we switch to production server after development and testing is done.

**Comment by samselikoff** [comment]
Mirage is bypassed if you run `ember serve --proxy` or build with `--environment production`.

This issue was about feeding your app's acceptance tests with data from the real backend.

**Comment by thatikuntap** [comment]
if run ember server, get below error, but it works once I remove http://localhost:8000/tools application adapter.

Error: Pretender will not respond to requests for URLs. It is not possible to accurately simluate the browser's CSP. Remove the http://localhost from http://localhost:8000/tools and try again

**Comment by thatikuntap** [comment]
sorry if I am posting this in wrong issue

**Comment by samselikoff** [comment]
yeah this is a different issue, move to #43

**Comment by JoshSmith** [comment]
What...what does this mean? :8ball:

**Comment by samselikoff** [comment]
Haha :) populating the fixture data with real backend data? Like vcr or proxy-fixtures? I'm not sure yet

**Comment by JoshSmith** [comment]
That would be all kinds of awesome. And make my life terribly easy. But sounds...hard.


What are the trade offs of the possible options?


—

**Comment by samselikoff** [comment]
I haven't thought through all this yet, there is a library that already does this: https://github.com/dockyard/ember-cli-proxy-fixtures though I'm not sure if it'd be possible to integrate. I'm still not sure if this is something Mirage should be responsible for

**Comment by samselikoff** [comment]
I have an idea for how to tackle this. I haven't thought through the implementation, but this is more or less the dream...

With Mirage factories, my acceptance tests look like this:

```js
test('should see title of a talk', function(assert) {
  server.create('talk', {title: 'Talk title'});
  visit('/talks');
  assertText('.Talks-table tr:first-child .Talks-table__title', 'Talk title');
});
```

The `server.create` line creates the talk in Mirage's database, so when the Ember app visits `/talks` and makes a request, the data is there to respond with.

Now, if you think about it, this test is saying, "given that this talk exists on the server, here's what the UI should look like." What if we introduced an adapter layer in Mirage that let it *actually* create the talk on the server? Meaning, this line

```js
server.create('talk')
```

would actually send an appropriate request to your running server (perhaps `localhost:3000`) to set up its database, and your Mirage route (when activated by your Ember app visiting `/talks`) would simply pass through to your running server? That way, the Mirage testing DSL serves both for working with your local Mirage server as well as your actual backend code. You could simply activate the switch, and all your acceptance tests would be running against your actual server.

**Comment by andrei1089** [comment]
@samselikoff  What's the status of this issue? The idea sounds really good, is there any work done on it?

**Comment by samselikoff** [comment]
No, unfortunately this is lower priority than factory relationships + serializers. The exploratory phase would essentially be, the flag for `passthrough` on all routes, and then having a flag so that `server.create` sends a POST request to your backend with the JSON payload from the factories. Then you'd have a route on your backend that could take the payload and essentially dump it into the database.

**Comment by lolmaus** [comment]
Hey @samselikoff, I'm so glad we're thinking [in the same plane](http://stackoverflow.com/questions/30402853/running-integration-acceptance-tests-on-the-frontend-need-an-api-for-the-fronte?noredirect=1#comment49502647_30402853).

I believe it's trivial to code a `server.create()` method on Ember side. The real work should be done on the backend and is framework-specific.

**Comment by samselikoff** [comment]
@lolmaus agreed but, getting something like a shared language for doing this in your ember app would be a great start. For example, if we were to incorporate this into Mirage, we could simply use the Mirage factories to generate the JSON, and pass that data to the backend to insert into the db.

So it might look like this:

1. `server.create('post')` is executed in an acceptance test
2. Mirage lookups the `PostFactory`, generates a record that looks like

    {title: 'Rails is omakase', author: 'DHH'}

3. Mirage takes that data and `POST`s to Ember CLI with some payload that looks like

    {
     resource: 'post',
     data: {title: 'Rails is omakase', author: 'DHH'}
    }
4. Ember CLI has a proxy in place that passes this request to `localhost:3000/mirage-testing`.
5. You could imagine a `mirage-testing` engine that the host rails app mounts onto `/mirage-testing`, which adds a controller that essentially knows how to CRUD server resources.
6. At this point, the data from the Mirage factory is in your rails backend, and as the acceptance test runs and makes a GET request to `/api/posts/1`, Mirage also passes this request to the real backend, which now responds with the same data your Ember acceptance test expects, and your tests can run.

**Comment by lolmaus** [comment]
I disagree.

You shouldn't reflect the structure of the backend's database in the frontend. That's tight coupling that will inevitably fall out of sync. I believe that the right way is to have fixture factories defined on the backend. Then they will always produce up to date entities.

The backend in test mode should expose an API for the frontend to request which models to create for each test and which factory defaults to override.

This API could look exactly like you suggest in `3`. But it will follow the backend's conventions which do not necessarily correspond to Ember's. They might differ a lot, differences being reconciled via adapters and serializers.

And the API could be invoked directly from tests via a simple helper:

```js
test('visiting /user-can-login-via-form', function(assert) {
  assert.expect(1);

  setupBackend({
    users: [
      alice: true,
      bob: {role: 'admin'}
    ]
  });

  andThen(function() {
    visit('/users');
  });

  andThen(function() {
    assert.equal(find('ul.users li').length, 2);
  });
});
```

So I basically imply that we don't need Mirage for this goal. I just don't see how Mirage can be useful for testing full stack.

**Comment by bdvholmes** [comment]
I think the thing of keeping fixture data in sync is a separate issue, but it may be possible to implement something that solves both problems.

I've been playing with setting up a proxy that sends requests that ember-cli-mirage doesn't know about to the backend - as opposed to the default proxy that you can set up with ember-cli. My aim is to automatically record responses from the backend to populate fixture data, but it could potentially help with this problem too.

**Comment by bdvholmes** [comment]
> I just don't see how Mirage can be useful for testing full stack.

If you only wanted to test full stack you wouldn't use Mirage. But once you've got your test suite written with Mirage it would be nice to be able to run it against a real api. @lolmaus is that what we're getting at here?

**Comment by samselikoff** [comment]
> You shouldn't reflect the structure of the backend's database in the frontend. That's tight coupling that will inevitably fall out of sync.

I agree with this, but I don't think `server.create('post', {title: 'Rails is Omakase'})` is reflecting db structure at all, it's reflecting *resource* structure, which your Ember app is absolutely aware of. The server could be storing the data as `name` instead of `title` but, from your Ember app's perspective, the test is semantic: *given* a post resource with a `title` attr of ____, the UI should look like ____.

**Comment by andrei1089** [comment]
I've been playing with adding a flag to make ```server.create``` also send data to the backend using the existing endpoints (POST to /<model_name>). 
Additionally, I've created an endpoint that cleans the backend database and I've made sure the ```server.create``` statements are in the correct order (they don't include IDs that are not yet created on the backend). 
Using the existing endpoints might not work all the time, but in my case as the permissions model is simple it was pretty straightforward to convert the existing acceptance tests to run against the real api. Here are my changes  https://github.com/andrei1089/ember-cli-mirage/commit/b58fc1c7e8251acb9917f625df2a6727f597a840, the code is pretty ugly as the url and the structure of the post data are hardcoded.
